File: app/src/main/java/com/example/simpleweathermvp/ui/home/HomePresenter.java
package com.example.simpleweathermvp.ui.home;

import android.view.View;

import com.example.simpleweathermvp.entity.City;
import com.example.simpleweathermvp.model.CityModel;

import java.util.List;

import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

public class HomePresenter {

    private CityModel citymodel;
    private HomeView homeView;
    public HomePresenter(CityModel model ) {
        this.citymodel = model;
    }

    void onViewCreated(HomeActivity homeView){
        this.homeView = homeView;
    }

    void onSuggvestionQwery(String qwery){
        citymodel.getCitySuggestion(qwery)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<List<City>>() {
                    @Override
                    public void onSubscribe(Disposable d) {

                    }

                    @Override
                    public void onNext(List<City> cities) {
                    homeView.showSuggvestionCities(cities);
                    }

                    @Override
                    public void onError(Throwable e) {

                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    void onSuggvestionCityClick(City city){

    }


}

File: app/src/main/java/com/example/simpleweathermvp/ui/home/HomePresenter.kt
package com.example.simpleweathermvp.ui.home

import android.annotation.SuppressLint
import android.util.Log
import com.example.simpleweathermvp.data.geteway.CityGateway
import com.example.simpleweathermvp.domain.city.*
import com.example.simpleweathermvp.domain.entity.City

import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.functions.Consumer
import io.reactivex.schedulers.Schedulers

class HomePresenter(
        private val homeView: HomeView,
        private val getCitySuggestionInteractor: GetCitySuggestionInteractor,
        private val addCityInteractor: AddCityInteractor,
        private val getAllCitiesInteractor: GetAllCitiesInteractor,
        private val removeCityInteractor: RemoveCityInteractor,
        private val getCityLocationInteractor: GetCityLocationInteractor

) {


    @SuppressLint("CheckResult")
    fun onSuggestionQuery(query: String){
        getCitySuggestionInteractor.execute(query)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({
                    homeView.showSuggvestionCities(it)
                }, {
                    Log.e("App", "Error", it)
                })
    }

    fun onSuggestionCitySelect(city: City) {
        addCityInteractor.execute(city)
    }

    @SuppressLint("CheckResult")
    fun getAllCities(){
        getAllCitiesInteractor.execute()
                .subscribe(Consumer {

                    homeView.showListCicies(it.map { cityTable ->
                        City(
                                id = cityTable.id,
                                name = cityTable.name,
                                country = cityTable.country,
                                lat = cityTable.lat,
                                lng = cityTable.lng
                        )
                    })
                })
    }

    fun removeCity(city: City){
        removeCityInteractor.execute(city)
    }

    @SuppressLint("CheckResult")
    fun clickOnCity(city: City) {

        if (city.lat != null && city.lng != null) {

            homeView.startWeatherActivity(city)

        } else {
            getCityLocationInteractor.execute(city)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({
                        addCityInteractor.execute(it)
                        homeView.startWeatherActivity(it)
                    }, {
                        Log.e("App", "Error", it)
                    })
        }
    }

}